src/poetry/vcs/git/backend.py

```diff
@@ -313,21 +314,47 @@ def _clone(cls, url: str, refspec: GitRefSpec, target: Path) -> Repo:
             if isinstance(e, AssertionError) and "Invalid object name" not in str(e):
                 raise
 
-            logger.debug(
-                "\nRequested ref (<c2>%s</c2>) was not fetched to local copy and"
-                " cannot be used. The following error was"
-                " raised:\n\n\t<warning>%s</>",
-                refspec.key,
-                e,
-            )
+            short_ref_found = False
 
-            raise PoetryConsoleError(
-                f"Failed to clone {url} at '{refspec.key}', verify ref exists on"
-                " remote."
-            )
+            if refspec.is_sha_short:
+                commit = cls._find_object_by_sha_prefix(local, refspec.key.encode())
+                if commit is not None:
+                    logger.debug(
+                        "\nResolved short SHA <c2>%s</c2> as <c2>%s</c2>",
+                        refspec.key,
+                        commit.id,
+                    )
+                    local.refs[b"HEAD"] = commit.id
+                    try:
```

Comment:
What is this try for? If the object was found in the local repo, why would it be missing here?

Reply:
Hmm, this is a good question. It probably was me being too cautious. Let me remove it.

Reply:
You've retrieved the commit object, so that will definitely be present. 

One of git's basic invariants (ensured by e.g. fetch, etc) is that if you have the commit object you also have the other objects it refers to*.

[*] This isn't strictly true under some circumstances, e.g. shallows - but that's not relevant here.